File: ios/Sources/RynBridgeShareKakao/KakaoTemplateMapper.swift
#if canImport(UIKit)
import Foundation
import RynBridge
import KakaoSDKTemplate

enum KakaoTemplateMapper {

    // MARK: - Feed Template

    static func feedTemplate(from payload: [String: AnyCodable]) throws -> FeedTemplate {
        guard let contentDict = payload["content"]?.dictionaryValue else {
            throw RynBridgeError(code: .invalidMessage, message: "Missing required field: content")
        }
        let content = try mapContent(from: contentDict)
        let social = payload["social"]?.dictionaryValue.flatMap { mapSocial(from: $0) }
        let buttons = mapButtons(from: payload["buttons"])
        return FeedTemplate(content: content, social: social, buttons: buttons)
    }

    // MARK: - Commerce Template

    static func commerceTemplate(from payload: [String: AnyCodable]) throws -> CommerceTemplate {
        guard let contentDict = payload["content"]?.dictionaryValue else {
            throw RynBridgeError(code: .invalidMessage, message: "Missing required field: content")
        }
        guard let commerceDict = payload["commerce"]?.dictionaryValue else {
            throw RynBridgeError(code: .invalidMessage, message: "Missing required field: commerce")
        }
        let content = try mapContent(from: contentDict)
        let commerce = mapCommerce(from: commerceDict)
        let buttons = mapButtons(from: payload["buttons"])
        return CommerceTemplate(content: content, commerce: commerce, buttons: buttons)
    }

    // MARK: - List Template

    static func listTemplate(from payload: [String: AnyCodable]) throws -> ListTemplate {
        guard let headerTitle = payload["headerTitle"]?.stringValue else {
            throw RynBridgeError(code: .invalidMessage, message: "Missing required field: headerTitle")
        }
        guard let headerLinkDict = payload["headerLink"]?.dictionaryValue else {
            throw RynBridgeError(code: .invalidMessage, message: "Missing required field: headerLink")
        }
        guard let contentsArray = payload["contents"]?.arrayValue else {
            throw RynBridgeError(code: .invalidMessage, message: "Missing required field: contents")
        }

        let headerLink = mapLink(from: headerLinkDict)
        let contents = try contentsArray.compactMap { item -> Content? in
            guard let dict = item.dictionaryValue else { return nil }
            return try mapContent(from: dict)
        }
        let buttons = mapButtons(from: payload["buttons"])
        return ListTemplate(headerTitle: headerTitle, headerLink: headerLink, contents: contents, buttons: buttons)
    }

    // MARK: - Shared Mappers

    static func mapContent(from dict: [String: AnyCodable]) throws -> Content {
        guard let title = dict["title"]?.stringValue else {
            throw RynBridgeError(code: .invalidMessage, message: "Missing required field: content.title")
        }
        guard let imageUrlString = dict["imageUrl"]?.stringValue,
              let imageUrl = URL(string: imageUrlString) else {
            throw RynBridgeError(code: .invalidMessage, message: "Missing or invalid field: content.imageUrl")
        }

        let linkDict = dict["link"]?.dictionaryValue ?? [:]
        let link = mapLink(from: linkDict)
        let description = dict["description"]?.stringValue
        let imageWidth = dict["imageWidth"]?.intValue
        let imageHeight = dict["imageHeight"]?.intValue

        return Content(
            title: title,
            imageUrl: imageUrl,
            imageWidth: imageWidth,
            imageHeight: imageHeight,
            description: description,
            link: link
        )
    }

    static func mapLink(from dict: [String: AnyCodable]) -> Link {
        let webUrl = dict["webUrl"]?.stringValue.flatMap { URL(string: $0) }
        let mobileWebUrl = dict["mobileWebUrl"]?.stringValue.flatMap { URL(string: $0) }
        let androidExecutionParams = mapStringDict(dict["androidExecutionParams"])
        let iosExecutionParams = mapStringDict(dict["iosExecutionParams"])
        return Link(
            webUrl: webUrl,
            mobileWebUrl: mobileWebUrl,
            androidExecutionParams: androidExecutionParams,
            iosExecutionParams: iosExecutionParams
        )
    }

    static func mapSocial(from dict: [String: AnyCodable]) -> Social? {
        Social(
            likeCount: dict["likeCount"]?.intValue,
            commentCount: dict["commentCount"]?.intValue,
            sharedCount: dict["sharedCount"]?.intValue,
            viewCount: dict["viewCount"]?.intValue,
            subscriberCount: dict["subscriberCount"]?.intValue
        )
    }

    static func mapCommerce(from dict: [String: AnyCodable]) -> CommerceDetail {
        return CommerceDetail(
            regularPrice: dict["regularPrice"]?.intValue ?? 0,
            discountPrice: dict["discountPrice"]?.intValue,
            discountRate: dict["discountRate"]?.intValue,
            fixedDiscountPrice: dict["fixedDiscountPrice"]?.intValue,
            productName: dict["productName"]?.stringValue,
            currencyUnit: dict["currencyUnit"]?.stringValue,
            currencyUnitPosition: dict["currencyUnitPosition"]?.intValue
        )
    }

    static func mapButtons(from value: AnyCodable?) -> [Button]? {
        guard let array = value?.arrayValue else { return nil }
        let buttons = array.compactMap { item -> Button? in
            guard let dict = item.dictionaryValue,
                  let title = dict["title"]?.stringValue else { return nil }
            let linkDict = dict["link"]?.dictionaryValue ?? [:]
            let link = mapLink(from: linkDict)
            return Button(title: title, link: link)
        }
        return buttons.isEmpty ? nil : buttons
    }

    static func mapStringDict(_ value: AnyCodable?) -> [String: String]? {
        guard let dict = value?.dictionaryValue else { return nil }
        var result: [String: String] = [:]
        for (key, val) in dict {
            if let str = val.stringValue {
                result[key] = str
            }
        }
        return result.isEmpty ? nil : result
    }
}
#endif

File: android/share-kakao/src/main/kotlin/com/devryner/rynbridge/share/kakao/KakaoTemplateMapper.kt
package com.devryner.rynbridge.share.kakao

import com.kakao.sdk.template.model.*
import com.devryner.rynbridge.core.BridgeValue
import com.devryner.rynbridge.core.RynBridgeError
import com.devryner.rynbridge.core.ErrorCode

object KakaoTemplateMapper {

    fun feedTemplate(from: Map<String, BridgeValue>): FeedTemplate {
        val contentMap = from["content"]?.dictionaryValue
            ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: content")
        val content = mapContent(contentMap)
        val social = from["social"]?.dictionaryValue?.let { mapSocial(it) }
        val buttons = mapButtons(from["buttons"])
        return FeedTemplate(content = content, social = social, buttons = buttons)
    }

    fun commerceTemplate(from: Map<String, BridgeValue>): CommerceTemplate {
        val contentMap = from["content"]?.dictionaryValue
            ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: content")
        val commerceMap = from["commerce"]?.dictionaryValue
            ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: commerce")
        val content = mapContent(contentMap)
        val commerce = mapCommerce(commerceMap)
        val buttons = mapButtons(from["buttons"])
        return CommerceTemplate(content = content, commerce = commerce, buttons = buttons)
    }

    fun listTemplate(from: Map<String, BridgeValue>): ListTemplate {
        val headerTitle = from["headerTitle"]?.stringValue
            ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: headerTitle")
        val headerLinkMap = from["headerLink"]?.dictionaryValue
            ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: headerLink")
        val contentsArray = from["contents"]?.arrayValue
            ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: contents")

        val headerLink = mapLink(headerLinkMap)
        val contents = contentsArray.mapNotNull { item ->
            item.dictionaryValue?.let { mapContent(it) }
        }
        val buttons = mapButtons(from["buttons"])
        return ListTemplate(headerTitle = headerTitle, headerLink = headerLink, contents = contents, buttons = buttons)
    }

    private fun mapContent(dict: Map<String, BridgeValue>): Content {
        val title = dict["title"]?.stringValue
            ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: content.title")
        val imageUrl = dict["imageUrl"]?.stringValue
            ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: content.imageUrl")
        val linkMap = dict["link"]?.dictionaryValue ?: emptyMap()
        val link = mapLink(linkMap)
        val description = dict["description"]?.stringValue
        val imageWidth = dict["imageWidth"]?.intValue?.toInt()
        val imageHeight = dict["imageHeight"]?.intValue?.toInt()

        return Content(
            title = title,
            imageUrl = imageUrl,
            imageWidth = imageWidth,
            imageHeight = imageHeight,
            description = description,
            link = link
        )
    }

    private fun mapLink(dict: Map<String, BridgeValue>): Link {
        return Link(
            webUrl = dict["webUrl"]?.stringValue,
            mobileWebUrl = dict["mobileWebUrl"]?.stringValue,
            androidExecutionParams = mapStringDict(dict["androidExecutionParams"]),
            iosExecutionParams = mapStringDict(dict["iosExecutionParams"])
        )
    }

    private fun mapSocial(dict: Map<String, BridgeValue>): Social {
        return Social(
            likeCount = dict["likeCount"]?.intValue?.toInt(),
            commentCount = dict["commentCount"]?.intValue?.toInt(),
            sharedCount = dict["sharedCount"]?.intValue?.toInt(),
            viewCount = dict["viewCount"]?.intValue?.toInt(),
            subscriberCount = dict["subscriberCount"]?.intValue?.toInt()
        )
    }

    private fun mapCommerce(dict: Map<String, BridgeValue>): Commerce {
        return Commerce(
            regularPrice = dict["regularPrice"]?.intValue?.toInt() ?: 0,
            discountPrice = dict["discountPrice"]?.intValue?.toInt(),
            discountRate = dict["discountRate"]?.intValue?.toInt(),
            fixedDiscountPrice = dict["fixedDiscountPrice"]?.intValue?.toInt(),
            productName = dict["productName"]?.stringValue,
            currencyUnit = dict["currencyUnit"]?.stringValue,
            currencyUnitPosition = dict["currencyUnitPosition"]?.intValue?.toInt()
        )
    }

    private fun mapButtons(value: BridgeValue?): List<Button>? {
        val array = value?.arrayValue ?: return null
        val buttons = array.mapNotNull { item ->
            val dict = item.dictionaryValue ?: return@mapNotNull null
            val title = dict["title"]?.stringValue ?: return@mapNotNull null
            val linkMap = dict["link"]?.dictionaryValue ?: emptyMap()
            val link = mapLink(linkMap)
            Button(title = title, link = link)
        }
        return buttons.ifEmpty { null }
    }

    fun mapStringDict(value: BridgeValue?): Map<String, String>? {
        val dict = value?.dictionaryValue ?: return null
        val result = mutableMapOf<String, String>()
        for ((key, v) in dict) {
            v.stringValue?.let { result[key] = it }
        }
        return result.ifEmpty { null }
    }
}
